syntax = "proto3";

message Tag {
    string key = 1;
    string value = 2;
}

message IRCClientMessage {
    repeated Tag tags = 1;
    string verb = 2;
    repeated string arguments = 3;
}

message IRCServerMessage {
    repeated Tag tags = 1;
    string prefix = 2;
    string verb = 3;
    repeated string arguments = 4;
}

message MessageFilter {
    repeated string verbs = 1;
}

message StreamRequest {
    MessageFilter filter = 1;
}

message SentResponse {}
message ConnectionRequest {}

message ConnectionResponse {
    bool result = 1;
}

service IRCConnection {
    rpc MessageStream(StreamRequest) returns (stream IRCServerMessage) {}
    rpc SendMessage(IRCClientMessage) returns (SentResponse) {}
    rpc DoConnection(ConnectionRequest) returns (ConnectionResponse) {}
}
